src/dynamic.cpp

```diff
@@ -500,3 +460,174 @@ void DYNAMIC::Search::print_result() const {
 
   std::cout << " nodes " << (totalCounter + counter);
 }
+
+namespace {
+
+    // Check if the position is semistatically unwinnable with recursive trivial progress.
+    bool is_unwinnable_with_trivial_progress(Position& pos, Color intendedWinner) {
+        MoveList<LEGAL> moveList(pos);
+
+        // Checkmate or Stalemate
+        if (moveList.size() == 0)
+            return !pos.checkers() || pos.side_to_move() == intendedWinner;
+
+        // Recursive trivial progress
+        if (moveList.size() == 1)
+        {
+            StateInfo stateInfo;
+            pos.do_move(*moveList.begin(), stateInfo);
+
+            // Returns true if the position is repeated.
+            bool res = stateInfo.repetition || is_unwinnable_with_trivial_progress(pos, intendedWinner);
+
+            pos.undo_move(*moveList.begin());
+            return res;
+        }
+
+        return SemiStatic::is_unwinnable(pos, intendedWinner);
+    }
+
+    bool side_to_move_can_capture_king(const Position& pos) {
+
+        for (const auto& m : MoveList<LEGAL>(pos)) {
+            if (!pos.empty(to_sq(m)) && type_of(pos.piece_on(to_sq(m))) == KING)
+                return true;
+        }
+        return false;
+    }
+
+    // Apply iterative deepening (find_mate may look deeper than maxDepth on
+    // rewarded variations)
+    DYNAMIC::SearchResult iterative_deepening(Position& pos, DYNAMIC::Search& search) {
+        for (int maxDepth = 2; maxDepth <= 1000; maxDepth++) {
+            // This choice seems empirically good
+            uint64_t limit = 10000;
+            search.set(maxDepth, search.actual_depth(), limit);
+            bool mate =
+                find_mate<DYNAMIC::FULL, DYNAMIC::ANY>(pos, search, 0, false, false);
+
+            if (!search.is_interrupted() && !mate)
+                search.set_unwinnable();
+
+            if (search.get_result() != DYNAMIC::UNDETERMINED ||
+                search.is_limit_reached())
+                break;
+        }
+
+        return search.get_result();
+    }
+}
+
+DYNAMIC::SearchResult DYNAMIC::full_analysis(Position& pos, DYNAMIC::Search& search) {
+    search.init();
+    search.set(0, 0, 0);
+
+    if (side_to_move_can_capture_king(pos)) {
+        search.set_unwinnable(); // is it Ok?
```

Comment:
Is this to handle illegal positions?
If so, I would not handle this case.

Reply:
I added the input position check because the current tool outputs incorrect mates in positions like the following:
`4K3/8/5R1k/8/8/8/8/8 w - - white`

Output:
`winnable f6f1 h6g5 f1a1 g5f4 a1b1 f4e4 b1a1 e4d5 a1b1 d5c6 b1a1 c6b7 a1b1 b7a8 e8d7 a8a7 d7c6 a7a6 b1a1# nodes 97 time 39 (4K3/8/5R1k/8/8/8/8/8 w - - white)`

However, if it is guaranteed that such positions cannot be submitted by the user, then the check can be removed.